### PR TITLE
Matrix Vector Multiplication

### PR DESCRIPTION
**Pull request by thezenelson**
## Purpose ##
This is primarily made of a module for calculating matrix vector multiplication. Along with this is an in-depth set of tests to confirm the functionality of the system.

## Commit 1 - Small Bug Fixes
* Changed the new hash to be ns based avoiding 2 threads created at the same time interfering
* Changed xcl2.cpp result string calculation to one that doesn't fail on my setup
* Added large allo uint type definitions

## Commit 2 - Matrix Vector Mult
* __init__ added mat_vec to KERNEL2SCHEDULE
* mat_vec.py
  * PE kernel which will output the dot products of every L elements of X and Y
  * mat_vec
    * X_load - takes horizontally packed X values and sends to corresponding stream
    * Y_load - saves all vector (Y) values on first iteration and sends them to all PE when needed
    * collect - Re integrates the results from the PE output streams in order
  * Schedule
    * Partition X, Y, and o so each PE has it's own input and output streams
    * Add pipeline pragma to load and store loops in addition to PE loops. The HLS compiler sometimes does this one automatically anyways.
    * Unfold PE elements to create each one individually
    * to pragmas to create streams for all communication with PE
* Testing
  *  PE tests for a large number of array lengths and "number of arrays to calculate"
  * Mat Vec tests including a large array of different sizes and number of PE
    * Includes a mix of software and csim tests

### CODE REVIEW

File: allo/harness/vitis/xcl2.cpp [hunk 1]
```diff
@@ -161,10 +161,11 @@ std::string convert_size(size_t size) {
...
 
-    std::stringstream stream;
-    stream << std::fixed << std::setprecision(2) << size_val;
-    std::string size_str = stream.str();
-    std::string result = size_str + " " + SIZES[div];
+    // std::stringstream stream;
+    // stream << std::fixed << std::setprecision(2) << size_val;
+    // std::string size_str = stream.str();
+    // std::string result = size_str + " " + SIZES[div];
+    std::string result = std::to_string(size_val) + " " + SIZES[div];
```
Comment (chhzh123) [inline allo/harness/vitis/xcl2.cpp]: What is this change used for?

File: allo/library/__init__.py [hunk 2]
```diff
@@ -8,8 +10,19 @@
...
+
+KERNEL2SCHEDULE.update(
+    {
+        systolic: schedule_systolic,
+        packed_systolic: schedule_systolic,
+        packed_int8xint8_systolic: schedule_systolic,
+    }
+)
+
+KERNEL2SCHEDULE[int8xint8_mat_vec] = schedule_int8xint8_mat_vec
```
Comment (chhzh123) [inline allo/library/__init__.py:28]: Why not combine this line with the previous dictionary construction?

File: allo/library/mat_vec.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+# pylint: disable=used-before-assignment, unsubscriptable-object, unused-import, unsupported-assignment-operation, missing-function-docstring
+
+from .. import dsl
+from ..ir.types import int8, int16, int32, index, Int, UInt
+from ..ir.utils import MockBuffer
+
+
+def pe_kernel[
+    L, D, P
+](x_in: "int8[P, D//P*L]", y_in: "int8[P, D//P*L]", o: "int8[P,D//P]", p: "index"):
```
Comment (chhzh123) [inline allo/library/mat_vec.py]: `o: "int8[P, D//P]"`

File: playground/mat_vec_test.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        mp.Process(target=test_basic_int8_gemvm_pe, args=args) for args in PE_tests
+    ]
+    for p in processes:
+        p.start()
+    for p in processes:
+        p.join()
+
+    # All full tests
+    # fmt: off
+    # pylint: fmt=off
```
Comment (chhzh123) [inline playground/mat_vec_test.py]: Is this pylint line needed?

**Review by chhzh123 — COMMENTED** [review]
Thanks for contributing to Allo! Most of the changes LGTM.

File: allo/library/__init__.py (hunk 2 above)
Comment (thezenelson) [reply allo/library/__init__.py:28]: The purpose of breaking the dictionary construction into separate parts, is to make it more modular with each addition corresponding to a separate .py file. I hope this helps with maintainability in the long term (making it easier to remove a library file.)

File: allo/library/mat_vec.py (hunk 3 above)
Comment (thezenelson) [reply allo/library/mat_vec.py]: I just pushed a fix for this.

File: playground/mat_vec_test.py (hunk 4 above)
Comment (thezenelson) [reply playground/mat_vec_test.py]: It's duplicative of the regular fmt: off line and I removed it.

File: allo/harness/vitis/xcl2.cpp (hunk 1 above)
Comment (thezenelson) [reply allo/harness/vitis/xcl2.cpp]: This was a fix for a cpp compiler error I was getting related to the stream. Though when I try and recreate it it works now. I'm just removing it for now.

**Review by thezenelson — COMMENTED** [review]
Thank you for spending the time to look over everything. What you mentioned should be fixed/responded to.

**Review by chhzh123 — COMMENTED** [review]
The new changes look fine. Two nits:
1. Remove `.pylintrc`. I suppose this was committed by accident.
2. Rename `mat_vec.py` to `gemv.py`. I will later rename `systolic.py` to `gemm.py`

**Comment by thezenelson** [comment]
1. `.pylintrc` was added on purpose so new setups will automatically use the configuration file. (I did remove it for now though.)
2. I changed this and the associated imports.

**Review by chhzh123 — APPROVED** [review]

**Pull request merged**